MoveIt! integration

adolfo-rt opened the pull request:
Unit tests build but are currently not passing.

Commit:
First prototype of motion planning support.

- Add dependencies on MoveIt!
- Add helper class that plans an approach trajectory when needed, and is able to
  reason about which planning groups to use
- Pending tasks documented as inline TODOs

Commit:
Refactor approach computation internals.

Planning group selection has been improved to select groups that:

- Span at least the joint of the input motion that change between current and
  goal configurations.

- Span at most all joints of the input motion.

Commit:
Proper support for setting waypoint vel, acc.

- Expose acceleration field through PlayMotion and MoveJointGroup.

- In MoveJointGroup, don't set zero velocity if unspecified. Since we now have
  the populateVelocities method, it's already being taken care of there.

Commit:
Action goal fails when approach computation fails.

- Fix for bug where goal remained active indefinitely.

Commit:
Log message aesthetics. Caps, better messages.

Commit:
Silence cppcheck warning.

- Type qualifiers ignored on function return type [-Wignored-qualifiers]:
  Function expecting const int returned int.

Commit:
Make planning optional. Deprecate reach_time.

- The action message has a new field: skip_planning, used to request for no
  motion planning to take place, case in which the approach time is
  automatically computed from a specified maximum velocity.

- The above point means that the reach_time parameter is no longer required,
  hence has been removed from the action goal message.

- A side-effect of this, is that the reach_time is computed by the approach
  planner, and no longer has to be forwarded all the way down to the
  MoveJointGroup instances.

Commit:
Fix crash when approach planner requisites not met

- If the required rosparam config is absent, print a descriptive error and
  don't crash.

Commit:
Make tests build.

- Adapt test logic to new message layout.

- Tests currently don't run successfully, as rrbot doesn't yet have a MoveIt!
  configuration.

Commit:
Isolate move_group async spinner from rest of node

Without this change, the synchronous spinning of play_motion was compromised.
Now play_motion preserves its synchronous spin behavior, and move_group
instances share a separate asynchronously serviced callback queue.

This changeset can be considered an initial fix for #31: When canceling a goal,
the cancel request will block until planning completes, but will then be
serviced correctly. The current planning latencies are small enough to make this
solution acceptable.

Commit:
Handle first waypoints with time_from_start == 0.

Two different scenarios have been addressed when the first waypoint has zero
time_from_start:

- Bugfix: If the approach trajectory is null, but non-planning joints move
  a reach time must be computed for the first waypoint.

- Cosmetic fix: Current state coincides with first waypoint. We make the
  time_from_start be a small, almost zero value to prevent the controllers
  from issuing a warning.

Commit:
Don't compute approach time if specified.

When skip_planning is set to True, the input motion might specify a valid
time_from_start for the first waypoint. The automatic reach time computation
is thus done only when the first waypoint contains a zero time_from_start.

Commit:
First iteration of move_joint script. Refs #33.

Commit:
Silence compiler warnings.

Commit:
Fix bug in applying computed reach time.

A regression was introduced in which the reach time was not propagated to all
motion waypoints.

Commit:
Simplify move_joint logic.

Commit:
Colorize move_joint output.

Commit:
Refactor move_joint file layout.

- Move guts to src/play_motion/move_joint.py

- scripts/move_joint is a thin wrapper that catches any unhandled
  exceptions, thus preventing lengthy tracebacks that confuse
  non-developers.

- Add move_joint to the install target.

po1 commented:
This means that the devel job will start to fail, which is not a problem as long as we fix these unit tests promptly. How long do you think it will take to adapt the unit tests to the new interface?

adolfo-rt commented:
One or two days of work. I need to create a rrbot moveit configuration and launch it. Then adapt existing tests a bit, i.e.,  review the checked conditions in some of them, and possibly force no planning.

The alternative for now would be to not build the tests altogether.

po1 commented:
Could we not just force no planning for the time being? Is the behavior so different?

adolfo-rt commented:
> Could we not just force no planning for the time being? Is the behavior so different?

Even if a particular motion is requested without planning, `play_motion` must be ready to take requests with planning. To do this, when constructing a `play_motion` server, we instantiate some moveit-specific members, which is slow (few seconds), hence can't be done on demand, for each request. As things are now, `play_motion` won't even initialize if the moveit configuration is absent.

po1 commented:
@adolfo-rt please update this branch with the possibility to work without a planner (only with `skip_planning: True`)
The required modifications to the existing unit tests should be fairly small. It would be great if we could not break them in this PR.

Commit:
Allow to disable motion planning altogether.

- Setting '~/approach_planner/disable_motion_planning=true' will cause
  play_motion to not initialize its motion planning instances.

- When deployed like this, the node will only accept goals expplicitly
  specifying 'skip_planning=true'

Commit:
Make tests build cleanly and pass.

- Add additional config to force no motion planning in existing tests.

- Fix compiler warnings.

- Lift test timeout. Otherwise as the test suite increases we'll need to
  keep on tweaking the value. Timeout errors are highly non-descriptive.

adolfo-rt commented:
Should be fine now. Units tests build cleanly and pass.

po1 commented on play_motion/src/play_motion/__init__.py:
```diff
@@ -0,0 +1 @@
+from .move_joint import move_joint, print_err, print_ok
```
In this specific case, people usually do this:

``` python
from .move_joint import *
```

adolfo-rt replied:
`.move_joint` implements other methods that I didn't want to directly expose, which is why I did the above. Is this considered a bad practice, or non-Pythonic?.

po1 replied:
I don't know what is the most pythonic way to do things. Usually people prefix functions, methods and members they  don't want to expose with an underscore.
That being said, since the list is small, it is fine as it is now here. If/when the list gets bigger, we'll reconsider.

po1 commented:
Also, a sample configuration file with comments, units and default values for `approach_planner` would be cool

adolfo-rt commented:
> Also, a sample configuration file with comments, units and default values for approach_planner would be cool

I could add it in the ros wiki doc. Sounds good?.

po1 commented:
> I could add it in the ros wiki doc. Sounds good?.

Since there already is a `sample.yaml`, I'd rather have it updated. That doesn't mean I don't want to put all of it in the wiki. It is just way simpler to copy a config file than clumsily copy and pasting from a wiki page.

adolfo-rt commented:
> How will this work if planning_disabled is not explicitly set and no MoveIt! configuration exists for the robot?

If the required `play_motion` part of the configuration is missing (e.g., which planning groups to use, and in which order), an exception is raised, and the node will die and issue a `ROS_FATAL` statement.

If the `play_motion` config is fine, but the MoveIt! side of things is absent, `play_motion` will not initialize. Currently it'll keep on waiting indefinitely for the `move_group` server to come up.

po1 commented:
This means that configurations that were working before will stop working after merging this. It is fine as long as we provide some documentation. I would add a line about this in the `README.md`

po1 commented:
@adolfo-rt I will merge this PR since the unit tests pass. For the little details, let's open an issue.

Commit:
Simplify how to disable motion planning.

- Move disable_motion_planning param to play_motion's namespace, instead of
  nesting it inside approach_planner.

- Update tests to exercise this behavior.

Commit:
Update README and sample config files

adolfo-rt commented:
Please review again

po1 commented:
Looks good, thank you very much for the effort. Merging.
